weave-js/src/components/Button/Button.tsx

```diff
@@ -4,153 +4,153 @@
   size?: ButtonSize;
   variant?: ButtonVariant;
   children?: ReactElement | string;
   active?: boolean;
   tooltip?: string;
   tooltipProps?: TooltipContentProps;
   twWrapperStyles?: React.CSSProperties;
+} & (IconButtonProps | LabelButtonProps);
+
+const sizeMap = {
+  small:
+    'gap-6 px-6 py-3 text-sm leading-[18px] [&_svg]:h-16 [&_svg]:w-16 h-24 w-24 p-0',
+  medium: 'gap-10 px-10 py-4 text-base [&_svg]:h-18 [&_svg]:w-18 h-32 w-32 p-0',
+  large: 'gap-8 px-12 py-8 text-base h-40 w-40 p-0',
 };
 
-export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
+const variantMap = {
+  primary: 'bg-teal-500 text-white hover:bg-teal-450',
+  secondary:
+    'bg-oblivion/[0.05] dark:bg-moonbeam/[0.05] text-moon-800 dark:text-moon-200 hover:bg-teal-300/[0.48] hover:text-teal-600 dark:hover:bg-teal-700/[0.48] dark:hover:text-teal-400',
+  ghost:
+    'bg-oblivion/[0.05] dark:bg-moonbeam/[0.05] text-moon-800 dark:text-moon-200 hover:bg-teal-300/[0.48] hover:text-teal-600 dark:hover:bg-teal-700/[0.48] dark:hover:text-teal-400',
+  quiet:
+    'text-moon-500 hover:text-moon-800 dark:hover:text-moon-200 hover:bg-oblivion/[0.05] dark:hover:bg-moonbeam/[0.05]',
+  destructive: 'bg-red-500 text-white hover:bg-red-450',
+};
+
+export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
   (
     {
       size = 'medium',
       variant = 'primary',
       icon,
-      startIcon = icon,
+      startIcon,
       endIcon,
       active,
       children,
       className = '',
       tooltip,
       tooltipProps,
-      twWrapperStyles = {},
+      twWrapperStyles,
       ...htmlAttributes
     },
     ref
   ) => {
-    const hasIcon = startIcon || endIcon;
-    if (!children && !hasIcon) {
-      console.error('Button: requires either children or an icon.');
-    }
-
-    const isSmall = size === 'small';
-    const isMedium = size === 'medium';
-    const isLarge = size === 'large';
-    const isPrimary = variant === 'primary';
-    const isSecondary = variant === 'secondary';
-    const isGhost = variant === 'ghost';
-    const isQuiet = variant === 'quiet';
-    const isDestructive = variant === 'destructive';
-
-    const hasBothIcons = startIcon && endIcon;
-    const hasOnlyOneIcon = hasIcon && !hasBothIcons;
-    const isIconOnly = hasOnlyOneIcon && !children;
-
     /**
      * The Tailwind wrapper is a div, which is a block-element. However, a normal button
      * wouldn't be wrapped and would be inline-block by default. These styles are a
      * necessary workaround to ensure Button is styled correctly despite the wrapper.
      */
-    const wrapperStyles = {
+    const wrapperStyles: CSSProperties = {
       display: 'inline-flex',
       width: className.includes('w-full') ? '100%' : undefined,
       ...twWrapperStyles,
     };
 
-    const button = (
+    /**
+     * Runtime validation of props' type constraints
+     */
+    const ButtonInner = () => {
```

Comment:
Enforces the type constraints above with helpful console output for invalid states.

Reply:
Sure, either work fine IMO but if `{buttonInner}` feels better than `<ButtonInner />` I can switch it over